bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/socket/RetryableSocket.java

```diff
@@ -0,0 +1,122 @@
+    }
+
+    /**
+     * Send a packet to the device, and expect a response.
+     * If retries in the thingConfig is > 0, we will send
+     * and receive repeatedly if we fail to get any response.
+     */
+    public byte @Nullable [] sendAndReceive(byte message[], String purpose) {
+        byte[] firstAttempt = sendAndReceiveOneTime(message, purpose);
+
+        if (firstAttempt != null) {
+            return firstAttempt;
+        }
+
+        if (thingConfig.getRetries() > 0) {
+            thingLogger.logTrace("Retrying sendAndReceive (for " + purpose + ") ONE time before giving up...");
+            return sendAndReceiveOneTime(message, purpose);
+        }
+
+        return null;
+    }
+
+    private byte @Nullable [] sendAndReceiveOneTime(byte message[], String purpose) {
+        // To avoid the possibility of a very fast response not being heard,
+        // we set up as much as possible for the response ahead of time:
+        byte response[] = new byte[1024];
+        DatagramPacket receivePacket = new DatagramPacket(response, response.length);
+        if (sendDatagram(message, purpose)) {
+            receivePacket = receiveDatagram(purpose, receivePacket);
+            if (receivePacket != null) {
+                return receivePacket.getData();
+            }
+        }
+
+        return null;
+    }
+
+    private boolean sendDatagram(byte message[], String purpose) {
+        try {
+            thingLogger
+                    .logTrace("Sending " + purpose + " to " + thingConfig.getIpAddress() + ":" + thingConfig.getPort());
+            if (socket == null || socket.isClosed()) {
+                thingLogger.logTrace("No existing socket ... creating");
+                socket = new DatagramSocket();
+                socket.setBroadcast(true);
+                socket.setReuseAddress(true);
+                socket.setSoTimeout(5000);
+            }
+            InetAddress host = InetAddress.getByName(thingConfig.getIpAddress());
+            int port = thingConfig.getPort();
+            DatagramPacket sendPacket = new DatagramPacket(message, message.length, new InetSocketAddress(host, port));
+            socket.send(sendPacket);
+            thingLogger.logTrace("Sending " + purpose + " complete");
+            return true;
+        } catch (IOException e) {
+            thingLogger.logError("IO error during UDP command sending " + purpose + " +: ", e);
+            return false;
+        }
+    }
+
+    private @Nullable DatagramPacket receiveDatagram(String purpose, DatagramPacket receivePacket) {
+        thingLogger.logTrace("Awaiting " + purpose + " response");
+
+        try {
+            if (socket == null) {
+                thingLogger.logError("receiveDatagram " + purpose + " for socket was unexpectedly null");
+            } else {
+                socket.receive(receivePacket);
+                thingLogger.logTrace("Received " + purpose + " (" + receivePacket.getLength() + " bytes)");
```

Comment:
Dont build a string like this for logging. When done this way it builds the string even when the logs are not in TRACE and costs the system some performance, see here...

https://www.openhab.org/docs/developer/guidelines.html#f-logging

logger.warn("Explain what went wrong and how to avoid it. You can have arguments {}.", someVariable, e);